rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/parametric/capabilities/DynamicCapabilities.java

```diff
@@ -0,0 +1,204 @@
+import org.rascalmpl.vscode.lsp.util.Lists;
+import org.rascalmpl.vscode.lsp.util.NamedThreadPool;
+import org.rascalmpl.vscode.lsp.util.concurrent.CompletableFutureUtils;
+
+/**
+ * Takes care of (un)registering capabilities dynamically.
+ * Receives a list of supported capabiltities ({@link AbstractDynamicCapability}) and notifies the client when capabilities change.
+ */
+public class DynamicCapabilities {
+
+    private static final Logger logger = LogManager.getLogger(DynamicCapabilities.class);
+
+    private final LanguageClient client;
+    private final Executor parallelExec;
+    private final Executor singleExec = NamedThreadPool.single("parametric-capabilities");
+    private final Collection<AbstractDynamicCapability<?>> supportedCapabilities;
+    private final Set<AbstractDynamicCapability<?>> staticCapabilities;
+    private final Map<String, Registration> currentRegistrations = new ConcurrentHashMap<>();
+
+    /**
+     * @param client The language client to send regiser/unregister requests to.
+     * @param exec The executor to use for asynchronous tasks.
+     * @param supportedCapabilities The capabilities to register with the client.
+     * @param clientCapabilities The capabilities of the client. Determine whether dynamic registration is supported at all.
+     */
+    public DynamicCapabilities(LanguageClient client, Executor exec, List<AbstractDynamicCapability<?>> supportedCapabilities, ClientCapabilities clientCapabilities) {
+        this.client = client;
+        this.parallelExec = exec;
+        this.supportedCapabilities = List.copyOf(supportedCapabilities);
+
+        // Check which capabilities to register statically
+        Set<AbstractDynamicCapability<?>> caps = new HashSet<>();
+        for (var cap : supportedCapabilities) {
+            if (cap.shouldRegisterStatically(clientCapabilities)) {
+                caps.add(cap);
+            }
+        }
+        // Set once and only read from now on
+        this.staticCapabilities = Collections.unmodifiableSet(caps);
+    }
+
+    /**
+     * Register static capabilities with the server.
+     */
+    public void registerStaticCapabilities(ServerCapabilities result) {
+        for (var cap : staticCapabilities) {
+            cap.registerStatically(result);
+        }
+    }
+
+    /**
+     * Update capabilities for language contributions.
+     * @param contribs The contributions to represent.
+     * @return A void future that completes when all capabilities are updated.
+     */
+    public CompletableFuture<Void> updateCapabilities(Collection<ILanguageContributions> contribs) {
+        // Copy the contributions so we know we are looking at a stable set of elements.
+        // If the contributions change, we expect our caller to call again.
+        var stableContribs = new LinkedHashSet<>(contribs);
+        return CompletableFutureUtils.reduce(supportedCapabilities.stream()
+            .filter(cap -> !staticCapabilities.contains(cap))
```

Comment:
if we're using equality, shouldn't the `Capability`  classes make sure to implement equals & hashcode? or do we really want reference equality here?